metricbeat/module/windows/perfmon/pdh_query_windows.go

```diff
@@ -0,0 +1,211 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build windows
+
+package perfmon
+
+import (
+	"regexp"
+	"strconv"
+	"syscall"
+	"unicode/utf16"
+	"unsafe"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	instanceNameRegexp = regexp.MustCompile(`.*\((.*)\).*`)
+	objectNameRegexp   = regexp.MustCompile(`(?:^\\\\[^\\]+\\|^\\)([^\\]+)`)
+)
+
+// Counter object will contain the handle and format of each performance counter.
+type Counter struct {
+	handle       PdhCounterHandle
+	format       PdhCounterFormat
+	instanceName string
+}
+
+// Query contains the pdh.
+type Query struct {
+	handle   PdhQueryHandle
+	counters map[string]*Counter
+}
+
+// CounterValue contains the performance counter values.
+type CounterValue struct {
+	Instance    string
+	Measurement interface{}
+	Err         error
+}
+
+// Open creates a new query.
+func (q *Query) Open() error {
+	h, err := PdhOpenQuery("", 0)
+	if err != nil {
+		return err
+	}
+	q.handle = h
+	q.counters = make(map[string]*Counter)
+	return nil
+}
+
+// AddCounter adds the specified counter to the query.
+func (q *Query) AddCounter(counterPath string, counter CounterConfig, wildcard bool) error {
+	if _, found := q.counters[counterPath]; found {
+		return errors.Errorf("Counter %s has been already added", counterPath)
+	}
+	var err error
+	var instanceName string
+	// Extract the instance name from the counterPath.
+	if wildcard || (!wildcard && counter.InstanceName == "") {
+		instanceName, err = matchInstanceName(counterPath)
+		if err != nil {
+			return err
+		}
+	} else {
+		instanceName = counter.InstanceName
+	}
+	h, err := PdhAddCounter(q.handle, counterPath, 0)
+	if err != nil {
+		return err
+	}
+
+	q.counters[counterPath] = &Counter{
+		handle:       h,
+		instanceName: instanceName,
+		format:       getFormat(counter.Format),
+	}
+	return nil
+}
+
+// CollectData collects the value for all counters in the query.
+func (q *Query) CollectData() error {
+	return PdhCollectQueryData(q.handle)
+}
+
+// GetFormattedCounterValues returns an array of formatted values for a query.
+func (q *Query) GetFormattedCounterValues() (map[string][]CounterValue, error) {
+	rtn := make(map[string][]CounterValue, len(q.counters))
+
+	for path, counter := range q.counters {
+		_, value, err := PdhGetFormattedCounterValue(counter.handle, counter.format|PdhFmtNoCap100)
+		if err != nil {
+			rtn[path] = append(rtn[path], CounterValue{Err: err})
```

Comment:
I was reading this line and 116 / 118 a few times. Is `rtn[path]` always set or could we run into a panic here?

Reply:
This shouldn't be a problem here, appending to a nil array works.